src/PAModel/Serializers/SourceSerializer.cs

```diff
@@ -950,5 +953,39 @@ private static void AddTestSuiteControl(CanvasDocument app, BlockNode controlIR)
             }
             app._screens[AppTestControlName].Children.Add(controlIR);
         }
+
+        /// <summary>
+        /// If the control name length is longer than 50, it is truncated and appended with a hash (to avoid collisions).
+        /// Checks the length of the escaped name, since its possible that the length is under 60 before escaping but goes beyond 60 later.
+        /// We do modulo by 1000 of the hash to limit it to 3 characters.
+        /// </summary>
+        /// <returns></returns>
+        private static string TruncateControlNameIfTooLong(string controlName)
+        {
+            var newControlName = Utilities.EscapeFilename(controlName);
+            if (newControlName.Length > MaxControlNameLength)
+            {
+                // limit the hash to 3 characters by doing a module by 1000
+                var hash = (GetHash(newControlName) % 1000).ToString("x3");
+                newControlName = newControlName.Substring(0, MaxControlNameLength - hash.Length) + hash;
+            }
+            return newControlName;
+        }
+
+        /// <summary>
+        /// djb2 algorithm to compute the hash of a string
+        /// </summary>
+        /// <param name="str"></param>
+        /// <returns></returns>
+        private static ulong GetHash(string str)
```

Comment:
>GetHash [](http://example.com/codeflow?start=29&length=7)

nit: 
addd comment "This must be determinstic and stable since it's used in file names". 
I'd add a unit test to verify that - hash a literal and compare to expected value (like what we do with checksums).